[schema] Support filtering individuals by last updated date

mafesan opened the pull request:
Now, individuals can be queried filtering by their last updated date.

The accepted filter formats are controlled by regular expressions matching two patterns:
* A comparison operator (>, >=, <, <=) and a date (e.g. `>=YYYY-MM-DDTHH:MM:SSZ`).
* A range operator (..) between two dates (e.g. `YYYY-MM-DDTHH:MM:SSZ..YYYY-MM-DDTHH:MM:SSZ`)

The accepted date format is ISO 8601, `YYYY-MM-DDTHH:MM:SSZ`, also accepting microseconds and time zone offset (`YYYY-MM-DDTHH:MM:SS.ms+HH:HH`).

Note: This PR is linked to issue #394 .

sduenas commented on sortinghat/core/schema.py:
```diff
@@ -77,6 +78,45 @@ def convert_json_field_to_generic_scalar(field, registry=None):
+        "date1": None,
+        "date2": None
+    }
+
+    # Filter with a comparison operator (>, >=, <, <=) and a date (e.g. `>=YYYY-MM-DDTHH:MM:SS`)
+    oper_comparison = r"^(<=?|>=?)(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})$"
```
This doesn't allow to look for an exact date. What would be the way to implement it?

mafesan replied:
As the accepted format I chose included date+time, it didn't make sense to me adding the `==` operator. I think the way it is done now, it could be set as a range filter like `YYYY-MM-DDT00:00:00..YYYY-MM-DDT23:59:59`, WDYT?

sduenas reviewed (commented):
I have some comments we need to clarify before we continue.

mafesan commented on sortinghat/core/schema.py:
```diff
@@ -77,6 +78,45 @@ def convert_json_field_to_generic_scalar(field, registry=None):
+        "date1": None,
+        "date2": None
+    }
+
+    # Filter with a comparison operator (>, >=, <, <=) and a date (e.g. `>=YYYY-MM-DDTHH:MM:SS`)
+    oper_comparison = r"^(<=?|>=?)(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})$"
```
You are right, but I chose to follow this format because it is the one that other date fields are using (like the enrollment dates in the `Enroll` mutation), as it is the one coming from the `DateTime` type from `graphene`.

sduenas commented on tests/test_schema.py:
```diff
@@ -1885,6 +1922,329 @@ def test_filter_non_exist_registry(self):
+        indv = individuals[0]
+        self.assertEqual(indv['mk'], 'a9b403e150dd4af8953a52a4bb841051e4b705d9')
+
+        self.assertEqual(indv['profile']['name'], None)
+        self.assertEqual(indv['profile']['email'], 'jsmith@example.com')
+        self.assertEqual(indv['profile']['isBot'], True)
+        self.assertEqual(indv['profile']['country']['code'], 'US')
+        self.assertEqual(indv['profile']['country']['name'], 'United States of America')
+
+        identities = indv['identities']
+        identities.sort(key=lambda x: x['uuid'])
+        self.assertEqual(len(identities), 3)
+
+        id1 = identities[0]
+        self.assertEqual(id1['email'], 'jsmith@example.com')
+
+        id2 = identities[1]
+        self.assertEqual(id2['email'], 'jsmith@bitergia.com')
+        self.assertEqual(id2['source'], 'scm')
+
+        id3 = identities[2]
+        self.assertEqual(id3['email'], 'jsmith@bitergia.com')
+        self.assertEqual(id3['source'], 'mls')
```
You can remove this. I don't think we need to check this. Checking the `mk` only should be enough.

sduenas reviewed (changes requested):
Please check my comments. The main problem is related to how to know when a filter is working and when it's not.

sduenas commented on sortinghat/core/schema.py:
```diff
@@ -77,6 +82,61 @@ def convert_json_field_to_generic_scalar(field, registry=None):
+    range_result = re.match(range_comparison, filter_value)
+
+    if not oper_result and not range_result:
+        raise InvalidFilterError(msg='Filter format is not valid')
```
It would be better if you explain what the problem is and how someone can fix it.

mafesan replied:
In this case I think it is better to include some basic documentation within the field itself using this graphene structure when the filter is declared:
```
 graphene.String(
        required=False,
        description='Filter with a comparison operator (>, >=, <, <=) and a date OR with a range operator (..) between two dates. E.g. `>=YYYY-MM-DDTHH:MM:SS`.'
    )
```
About the information the error has to provide, the problem there is the filter did not match with any of the accepted regular expressions. Should this error print the not matching parts together with the expression that should match?

mafesan replied:
This description would look like this in the API web documentation:

![doc-last-updated-filter](https://user-images.githubusercontent.com/9032790/100731172-52797000-33cb-11eb-8d26-a337716ac066.png)

@sduenas, WDYT ?

sduenas replied:
I think it's good to have that in the documentation. However, one of the problems of the exception is that it doesn't says what filter is failing, so the user won't be able to know what to change.

mafesan replied:
Oh, I see what you meant. The error was not telling `last_updated` is the filter field that is failing, I've just updated it catching the exception when the filter fails while the query is being resolved.

mafesan replied:
BTW, I kept the piece of documentation I posted above with some real examples:
* `>=2020-10-12T09:35:06.13045+01:00` 
* `2020-10-12T00:00:00..2020-11-22T00:00:00`

sduenas commented on sortinghat/core/schema.py:
```diff
@@ -77,6 +82,61 @@ def convert_json_field_to_generic_scalar(field, registry=None):
+            raise InvalidFilterError(msg=e)
+
+        if filter_data['date1'] > filter_data['date2']:
+            raise InvalidFilterError(msg='Date range is invalid, date 2 must be greater than date 1')
```
Instead of saying `date 2` and `date 1`, why don't we say `Date range is invalid. Upper bound must be greater than the lower bound", or `Date range is invalid. First date must be greater than the second one". Let's try to find something clear.

sduenas reviewed (changes requested):
LGTM but some minor changes are needed.

mafesan commented:
Thank you so much for your comments @sduenas ! Now, ` parse_date_filter` method raises `ValueError` when the filter format is wrong or the date range is invalid; and `InvalidDateError` when the filter contains a wrong date. Both errors are captured when the query is being resolved, raising a `InvalidFilterError`, which gives information about the filter causing the problem and the cause of the error.

sduenas reviewed (changes requested):
The code looks ok. I've been testing it but it doesn't returns what it should, though.
For example, see the two screenshots. The first one shows the individuals that I have on the database. The second one shows a query that should return those individuals.

![imagen](https://user-images.githubusercontent.com/833352/100775064-0c400300-3403-11eb-9caa-752d895b8e0f.png)

![imagen](https://user-images.githubusercontent.com/833352/100775114-1cf07900-3403-11eb-821b-a2b5ce8d337a.png)

mafesan commented:
> The code looks ok. I've been testing it but it doesn't returns what it should, though.
> For example, see the two screenshots. The first one shows the individuals that I have on the database. The second one shows a query that should return those individuals.
> 
> ![imagen](https://user-images.githubusercontent.com/833352/100775064-0c400300-3403-11eb-9caa-752d895b8e0f.png)
> 
> ![imagen](https://user-images.githubusercontent.com/833352/100775114-1cf07900-3403-11eb-821b-a2b5ce8d337a.png)

You are right @sduenas, thank you for the heads-up. At least one of the individuals should be returned when filtering by that date... :thinking: 
I did test it and it worked, as I added the `__date__` section when filtering, as they explain in [Django documentation](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#date), but thinking about it, I added this part when I was taking the date as a string, and then Django casted that as a datetime object.

Now, the `str_to_datetime` method from GrimoireELK does this conversion already in the `parse_date_filter` method, so instead of calling the filter as `<filed_name>__date__<operator>` (`last_modified__date__gte`), it can be rewritten directly as `<field_name>__<operator>` (`last_modified__gte`), as the date is already a datetime object. After applying this change, it seems it works as expected (the unit tests worked in both cases, though).

Please, can you test again with the new changes? Thanks!

mafesan commented:
@sduenas I have modified the unit tests to cover the case you mentioned. 

The problem was related about how to build the Django query filter:  when building the query as `<filed_name>__date__<operator>`, the functionality was not working from the interface but the unit tests did not fail.

To fix this error, I modified how the Django query filter was built to `<field_name>__<operator>` but the tests kept working with no modification. Now I have added more cases to ensure this case is covered: the tests work with this updated filter and they fail if the filter is replaced by the former one.

evamillan commented:
@mafesan The JavaScript `toISOString` function returns a date with a `YYYY-MM-DDTHH:mm:ss.sssZ` format, but this kind of date raises a `Filter format is not valid` error when I test the PR. This is the format that we use for all the other dates, like the ones in the `enroll` mutation. Could you please change it?

mafesan commented:
> @mafesan The JavaScript `toISOString` function returns a date with a `YYYY-MM-DDTHH:mm:ss.sssZ` format, but this kind of date raises a `Filter format is not valid` error when I test the PR. This is the format that we use for all the other dates, like the ones in the `enroll` mutation. Could you please change it?

Thank you for the heads-up @evamillan! I have modified the accepted date formats to include the format you mentioned, I have tested it with the search box from the changes in #408 and it worked for me, can you please confirm it works for you too? Thanks!

evamillan commented:
I tested it with the last changes and it works. Thank you @mafesan !

sduenas reviewed (approved):
LGTM
